ncmcm/bundlenet/bundlenet.py

```diff
@@ -34,29 +47,38 @@ class BunDLeNet(Model):
             discrete behavioural states
             For continuous-valued behaviours, this stipulates the number of
             behaviour variables
+
+        input_shape (tuple):
+            Shape of the input data.
+
     """
 
-    def __init__(self, latent_dim: int, num_behaviour: int):
+    def __init__(self, latent_dim: int, num_behaviour: int, input_shape: tuple):
         super(BunDLeNet, self).__init__()
+        in_features = np.prod(input_shape[-2:])
         self.latent_dim = latent_dim
         self.num_behaviour = num_behaviour
-        self.tau = tf.keras.Sequential([
-            layers.Flatten(),
-            layers.Dense(50, activation='relu'),
-            layers.Dense(30, activation='relu'),
-            layers.Dense(25, activation='relu'),
-            layers.Dense(10, activation='relu'),
-            layers.Dense(latent_dim, activation='linear'),
-            layers.GaussianNoise(0.05)
-        ])
-        self.T_Y = tf.keras.Sequential([
-            layers.Dense(latent_dim, activation='linear'),
-        ])
-        self.predictor = tf.keras.Sequential([
-            layers.Dense(num_behaviour, activation='linear')
-        ])
-
-    def call(self, x):
+        self.tau = nn.Sequential(
+            nn.Flatten(),
+            nn.Linear(in_features, 50),
+            nn.ReLU(),
+            nn.Linear(50, 30),
+            nn.ReLU(),
+            nn.Linear(30, 25),
+            nn.ReLU(),
+            nn.Linear(25, 10),
+            nn.ReLU(),
+            nn.Linear(10, latent_dim),
+            GaussianNoise(0.05),
+        )
```

Comment:
Good job finding out that the normalization layer did not do anything! I verified that as well. Could you think of any replacement that results in the latent space being normalised? Basically, we don't want the method to 'cheat' by learning a manifold that is so small that it leads to a low dynamical loss even though it does not contain predictive information.

Reply:
I have tried both using layer normalization _LayerNorm_ as well as batch normalization _BatchNorm1d_
For each, I have tried adding them:
• after every linear layer in tau
• after every linear layer in tau and after the linear layer of T_Y
• only after the last linear layer in tau
• only after the last linear layer in tau and after the linear layer of T_Y

The layer normalization didn’t work for me in all cases, it would just create a line for the continuous model/dmeo and a flat circle for the discrete one.
Batch normalization generally only worked similarly to the current version when placed only after the last linear layer of tau. So I updated the code and demos with _BatchNorm1d_ after the last linear layer of tau.